[AArch64] Add MATCH loops to LoopIdiomVectorizePass

rj-jesus opened the pull request:
This patch adds a new loop to LoopIdiomVectorizePass, enabling it to
recognise and vectorise loops such as:
```cpp
template<class InputIt, class ForwardIt>
InputIt find_first_of(InputIt first, InputIt last,
                      ForwardIt s_first, ForwardIt s_last)
{
  for (; first != last; ++first)
    for (ForwardIt it = s_first; it != s_last; ++it)
      if (*first == *it)
        return first;
  return last;
}
```

These loops match the C++ standard library function std::find_first_of.

The loops are vectorised using `@experimental.vector.match` in #101974.

rj-jesus committed:
[AArch64] Add MATCH loops to LoopIdiomVectorizePass

This patch adds a new loop to LoopIdiomVectorizePass, enabling it to
recognise and use @llvm.experimental.vector.match to vectorise loops
such as:

    char* find_first_of(char *first, char *last,
                        char *s_first, char *s_last) {
      for (; first != last; ++first)
        for (char *it = s_first; it != s_last; ++it)
          if (*first == *it)
            return first;
      return last;
    }

These loops match the C++ standard library's std::find_first_of.

rj-jesus commented:
Hi @david-arm and @paulwalker-arm, I've rebased this patch to use the version of `@llvm.experimental.vector.match` committed last week. I've also added a simple cost model which currently only lets the transformation kick in for supported SVE vectors.

Currently, a simple double find loop like the one in the description should result in something like:
```gas
find_first_of_i8:                       // @find_first_of_i8
	.cfi_startproc
// %bb.0:
	cmp	x0, x1
	b.eq	.LBB0_7
// %bb.1:
	cmp	x2, x3
	b.eq	.LBB0_7
// %bb.2:                               // %.preheader
	ptrue	p0.b, vl16
.LBB0_3:                                // =>This Loop Header: Depth=1
                                        //     Child Loop BB0_4 Depth 2
	whilelo	p1.b, x0, x1
	mov	x8, x2
	mov	x9, x2
	and	p1.b, p0/z, p0.b, p1.b
	ld1b	{ z0.b }, p1/z, [x0]
.LBB0_4:                                //   Parent Loop BB0_3 Depth=1
                                        // =>  This Inner Loop Header: Depth=2
	whilelo	p2.b, x8, x3
	and	p2.b, p0/z, p0.b, p2.b
	ld1b	{ z1.b }, p2/z, [x9]
	mov	z2.b, b1
	sel	z1.b, p2, z1.b, z2.b
	mov	z1.q, q1
	match	p2.b, p1/z, z0.b, z1.b
	b.ne	.LBB0_8
// %bb.5:                               //   in Loop: Header=BB0_4 Depth=2
	add	x9, x9, #16
	add	x8, x8, #16
	cmp	x9, x3
	b.lo	.LBB0_4
// %bb.6:                               //   in Loop: Header=BB0_3 Depth=1
	add	x0, x0, #16
	cmp	x0, x1
	b.lo	.LBB0_3
.LBB0_7:                                // %.loopexit1
	mov	x0, x1
	ret
.LBB0_8:                                // %.loopexit
	ptrue	p0.b
	brkb	p0.b, p0/z, p2.b
	incp	x0, p0.b
	ret
```

Could you please let me know if you have any thoughts or suggestions? Many thanks in advance!

sjoerdmeijer commented:
The output looks cool. Naive drive-by question: is it faster than what we currently generate? :)

rj-jesus commented:
It is, in the table below I've summarised the speedups obtained on a Neoverse V2 for a combination of search (columns) and needle (rows) vector sizes for the loop in the PR's description:
![data](https://github.com/user-attachments/assets/5aca75da-3453-4790-80f9-a463ad946f5d)
For item (4, 80) in the matrix, for example, we have a speedup of 16.1x, which means that, for search arrays of 80 characters and needle arrays of 4 characters, where the match happens at the last element of both arrays, we get an average speedup of 16.1 over the current scalar loops. I can probably share the benchmark I used to collect these numbers if you think that would be useful, I would just need to double-check this internally.

Overall the vectorised loops are at least competitive with the scalar loops (and often obtain substantial speedups over the latter). The exceptions are for degenerate cases of search or needle arrays consisting of a single element (top left corner of the matrix).

The output we generate for the vectorised loops also contains four "unnecessary instructions", in particular:
* `mov	x8, x2` - redundant copy of `mov x9, x2`
* `add	x8, x8, #16` - as above, copy of `add x9, x9, #16`
* `mov	z1.q, q1` - we do this for correctness to broadcast the needles from a segment to a full SVE vector, but this is not necessary here since we effectively only work with the first 128-bit segment of the SVE register
* `ptrue	p0.b` - this is introduced by `experimental.cttz.elts` but is technically not needed in this case since we could reuse the previously defined `p0`

None of these issues are related to the transformation itself (and don't affect its performance that much anyway), so I think they can be addressed separately.

sjoerdmeijer commented:
Thanks for the experiments and data, I think that looks really good! This always looks like a win, except when the size is or 1 or 2. That is most likely so exceptional we don't have to consider runtime size checks to either fall into the scalar or the vector loop.

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+      ScalableVectorType::get(Type::getInt1Ty(Ctx), VF)};
+  IntrinsicCostAttributes Attrs(Intrinsic::experimental_vector_match, Args[2],
+                                Args);
+  if (TTI->getIntrinsicInstrCost(Attrs, TTI::TCK_SizeAndLatency) > 4)
```
This looks a bit out of place, most of the code here deals with checking the structure of the CFG. This is essentially checking whether the intrinsic is supported. Can we give this a different home, can we do this earlier?

rj-jesus replied:
We need to do this here or later since this is when we match the `CharTy` of the loads, which we use to evaluate the cost of the intrinsic. Would it make sense to move it later?

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+      MatchPred != ICmpInst::Predicate::ICMP_EQ)
+    return false;
+
+  LLVM_DEBUG(dbgs() << "FOUND IDIOM IN LOOP: \n" << *CurLoop << "\n\n");
```
Nit: maybe less capitals in the debug message.

Actually, I wouldn't mind some more debug messages, especially when it is rejected.

You're always returning true, maybe at least reports when it is rejected, something like:

```
   if(!recognizeFindFirstByte()) {
       LLVM_DEBUG("...");
       return;
   }
  transformFindFirstByte();
```

and preferably also a reason why it is rejected.

rj-jesus replied:
Thanks, I think this was from `LoopIdiomVectorize::recognizeByteCompare` but I've formatted it properly now.

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  BasicBlock *BB5 = BasicBlock::Create(Ctx, "", SPH->getParent(), SPH);
+
+  // Update LoopInfo with the new loops.
+  auto OL = LI->AllocateLoop();
```
Nit: I would prefer OuterLoop / InnerLoop

rj-jesus replied:
Thanks, done.

sjoerdmeijer commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+
+  // (1) Load a vector's worth of A and branch to the inner loop.
+  Builder.SetInsertPoint(BB1);
+  PHINode *PNA = Builder.CreatePHI(PtrTy, 2, "pa");
```
I wouldn't mind a slightly longer but more descriptive name, e.g. psearch or just search.

rj-jesus replied:
Thanks, done.

rj-jesus commented:
Hi @sjoerdmeijer, thanks very much for the feedback! I'll update the patch later this week.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -914,6 +914,23 @@ AArch64TTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+    InstructionCost Cost = 4;
+    // For fixed-length vectors we currently need an extra five--six
+    // instructions besides the MATCH.
+    if (isa<FixedVectorType>(RetTy))
```
The intrinsic itself supports some types of fixed-length vectors, which is why I added this cost here. Nevertheless, since we don't use these in the transformation right now, I've marked their cost as invalid until we have a use case for them.

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -79,6 +104,12 @@ static cl::opt<unsigned>
+static cl::opt<bool>
+    DisableFindFirstByte("disable-loop-idiom-vectorize-find-first-byte",
+                         cl::Hidden, cl::init(false),
+                         cl::desc("Proceed with Loop Idiom Vectorize Pass, but "
```
Thanks, done.

rj-jesus commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
```diff
@@ -0,0 +1,123 @@
+;   }
+define ptr @find_first_of_i8(ptr %0, ptr %1, ptr %2, ptr %3) #0 {
+; SVE2-LABEL: define ptr @find_first_of_i8(
+; SVE2:         {{%.*}} = call <vscale x 16 x i1> @llvm.experimental.vector.match.nxv16i8.v16i8(<vscale x 16 x i8> {{%.*}}, <16 x i8> {{%.*}}, <vscale x 16 x i1> {{%.*}})
```
Thanks, done. I've also removed the two run lines (replaced by one) to avoid duplicate negative test matching.

rj-jesus commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
```diff
@@ -0,0 +1,123 @@
+; RUN: opt -mattr=+sve2 -mtriple=aarch64 -passes='loop(loop-idiom-vectorize)' -S < %s | FileCheck -check-prefix=SVE2 %s
+; RUN: opt -mattr=-sve2 -mtriple=aarch64 -passes='loop(loop-idiom-vectorize)' -S < %s | FileCheck -check-prefix=NOSVE2 %s
+
```
Thanks, done.

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // We are expecting the blocks below. For now, we will bail out for almost
+  // anything other than this.
+  //
+  // Header:
```
Thanks, done.

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+      LoopBlocks[3]->sizeWithoutDebug() > 3)
+    return false;
+
+  // Check that no instruction other than IndPhi has outside uses.
```
It wasn't but is now, thanks.

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // and decide if we should use it.
+  // Note: VF could be parameterised, but 128-bit vectors sounds like a good
+  // default choice for the time being.
+  unsigned VF = 128 / CharTy->getIntegerBitWidth();
```
It should also work for `i16` (which MATCH also supports). The intrinsic cost is going to return invalid for other cases (and bail out).

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,361 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // We will have the following loops:
+  // (O) Outer loop where we iterate over the elements of the search array (A).
+  // (I) Inner loop where we iterate over the elements of the needle array (B).
+  //
```
Thanks, I've renamed these to Search/Needle. Please have a look and let me know if that seems okay now.

rj-jesus commented:
Thanks again for your feedback @sjoerdmeijer, I believe I've addressed most of your comments in my latest commit.
Please let me know if I've missed anything or if you have any other comments!

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -914,6 +914,23 @@ AArch64TTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
     }
     break;
   }
+  case Intrinsic::experimental_vector_match: {
```
Could you pull the cost model changes out into a separate PR please? I think they make sense as a standalone change. Also, it would be good to add some tests in llvm/test/Analysis/CostModel/AArch64/sve-intrinsics.ll. You can look at `@reductions` as an example.

rj-jesus replied:
Thanks, I've done so in #118512. Please let me know if you have any other comments!

sjoerdmeijer commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
```diff
@@ -0,0 +1,429 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -mtriple=aarch64 -mattr=+sve -passes='loop(loop-idiom-vectorize)' -verify-loop-info -verify-dom-info -S < %s | FileCheck %s
```
Add a RUN line to check option '-disable-loop-idiom-vectorize-find-first-byte'

sjoerdmeijer reviewed (approved):
This looks good to me.

rj-jesus committed:
Add RUN line with -disable(...) and refactor tests

rj-jesus commented:
> This looks good to me.

Many thanks, @sjoerdmeijer! I've just made the changes you suggested and rebased this on 2fe30bc6693c60d76c7e44d9fd6323c39125c19e, please let me know if you have any other comments.

Also, if anyone else is planning to have a look at this, could you please give me a shout?
Otherwise I'll merge in the next couple of days.

Thanks again!

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,359 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // We expect outside uses of `IndPhi' in ExitSucc (and only there).
+  for (User *U : IndPhi->users())
+    if (!CurLoop->contains(cast<Instruction>(U)))
+      if (auto *PN = dyn_cast<PHINode>(U); !PN || PN->getParent() != ExitSucc)
```
nit: I think I prefer declaring the `PN` variable separately to avoid the multi-statement if line, i.e.

```
  for (User *U : IndPhi->users())
    if (!CurLoop->contains(cast<Instruction>(U))) {
      auto *PN = dyn_cast<PHINode>(U);
      if (!PN || PN->getParent() != ExitSucc)
        return false;
    }
```

I personally find it easier to read.

rj-jesus replied:
Thanks, done.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,359 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // One PHI comes from the outer loop (PNA), the other one from the inner loop
+  // (PNB). PNA effectively corresponds to IndPhi.
+  if (InnerLoop->contains(PNA))
+    std::swap(PNA, PNB);
```
nit: Is it worth renaming PNA and PNB to PNOuter and PNInner to make it more obvious what A and B are expected to be?

Might be good to do something similar for StartA, IndexA, GEPA, etc.

rj-jesus replied:
Thanks, I've renamed them to Search/Needle as appropriate as I had already done in expand/transformFindFirstByte.

david-arm commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
```diff
@@ -0,0 +1,416 @@
+;   char* find_first_of(char *first, char *last, char *s_first, char *s_last) {
+;     for (; first != last; ++first)
+;       for (char *it = s_first; it != s_last; ++it)
+;         if (*first == *it)
```
Given this can exit early before loading the rest of the needle (`*it`) you need memory checks to ensure `*it` does not cross a page boundary, similar to what we did for the mismatch expand code.

rj-jesus replied:
Thanks very much for catching this - I've added these checks. Could you please have a look and let me know if it looks better now?

david-arm reviewed (commented):
The patch is looking mostly good. I've still got a bit more to review, but at the moment there is a serious bug with this so can you hold off landing this until it's fixed? The inner loop can exit early before you've loaded the rest of the needle so it needs memory checks before entering the vector loop to ensure the needle doesn't cross a page boundary.

david-arm commented:
In fact, I think we need page boundary checks for the pointer in the outer loop as well. Suppose you find a match in your needle after only two iterations in your scalar outer loop? Your third iteration could cross a page boundary and seg fault. The scalar version of the loop will be fine because we exited before the fault, but the vector version will crash.

david-arm commented:
The memory checks for the inner loop only need to be done once I assume, since I'd expect the start value for the needle to be loop invariant? If not, you'll need to calculate the entire needle range for the whole outer loop. I'd expect a minor hit to performance, but given the fantastic improvements you're reporting I'd still expect it to be a nice win. :) For the mismatch case the trip counts were typically quite small and the memory checks reduced the performance from around 8% -> 7% or something like that.

rj-jesus commented:
> The patch is looking mostly good. I've still got a bit more to review, but at the moment there is a serious bug with this so can you hold off landing this until it's fixed?

Of course, I will do, thank you very much for the heads up!

> The inner loop can exit early before you've loaded the rest of the needle so it needs memory checks before entering the vector loop to ensure the needle doesn't cross a page boundary.

> In fact, I think we need page boundary checks for the pointer in the outer loop as well. Suppose you find a match in your needle after only two iterations in your scalar outer loop? Your third iteration could cross a page boundary and seg fault. The scalar version of the loop will be fine because we exited before the fault, but the vector version will crash.

I think I'm missing something. We should never be reading beyond the end of the search or needle arrays (if we do, that's definitely a bug), so I don't think we would be prone to segfaulting. The idea was to use `llvm.get.active.lane.mask` to create a valid predicate for the loads, shouldn't that work?

david-arm commented:
> I think I'm missing something. We should never be reading beyond the end of the search or needle arrays (if we do, that's definitely a bug), so I don't think we would be prone to segfaulting. The idea was to use llvm.get.active.lane.mask to create a valid predicate for the loads, shouldn't that work?

Doesn't your predicate simply ensure you never load more data than the range of the loop, but it doesn't protect you against seg faults? It is possible to use masks to avoid seg faults, but you'd have to split your loops up page by page which is pretty complicated. Consider one of the example C codes in your test:

```
;   char* find_first_of(char *first, char *last, char *s_first, char *s_last) {
;     for (; first != last; ++first)
;       for (char *it = s_first; it != s_last; ++it)
;         if (*first == *it)
;           return first;
;     return last;
;   }
```

The scalar loop may find a match `if (*first == *it)` on the first iteration of your inner loop. Suppose the pointer `it` starts at the last byte of a page and (s_last - s_first) = 16. In the scalar version you'll exit the loop before cross the page boundary so you won't seg fault, but in the vector version you'll load all 16 bytes which cross a page boundary. If the second page isn't mapped in memory the version will seg fault. This is the fundamental problem with vectorising loops that contain early exits.

david-arm commented:
Alternatively, if you can prove at compile time the vector loop won't fault you can also do that as a first step, then have a follow-on patch that enables loop versioning. For example, if you have some IR where the pointers all come from an `alloca` and you know the trip counts for both loops you can use `llvm::isDereferenceableAndAlignedInLoop` to ask the question if the load pointer can be unconditionally dereferenced.

rj-jesus committed:
Add page boundary checks and address other comments

rj-jesus commented:
Many thanks for your comments, @david-arm, I should have addressed them in the latest revision of the patch and added runtime page boundary checks as you had pointed out. Could you please let me know if you have any other suggestions?

rj-jesus commented:
Ping :)

appujee commented:
@david-arm do you have any pending comments or we can merge it?

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,400 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  Value *ISearchStart = Builder.CreatePtrToInt(SearchStart, I64Ty);
+  Value *ISearchEnd = Builder.CreatePtrToInt(SearchEnd, I64Ty);
+  Value *INeedleStart = Builder.CreatePtrToInt(NeedleStart, I64Ty);
+  Value *INeedleEnd = Builder.CreatePtrToInt(NeedleEnd, I64Ty);
```
Is this guaranteed to be safe? For example, do you know that `NeedleEnd` is invariant in the outer loop? Perhaps I didn't look closely enough, but I didn't see anything in the code that checks for outer loop invariance.

david-arm replied:
Maybe you get this check for free due to how you've checked all the IR in the outer loop.

rj-jesus replied:
I think we do get the check for free due to how the IR is matched, but I've added checks for loop invariance [here](https://github.com/llvm/llvm-project/pull/101976/files#diff-cc41ba3d3c42df08ea84e2c280cc1060bf3c91f6fadf225fecb7334c74267080R1161-R1164) in any case. Is that what you had in mind?

rj-jesus committed:
Move tests to llvm/test/Transforms/LoopIdiom/AArch64

rj-jesus committed:
Add names to blocks and variables

rj-jesus committed:
Add checks for loop invariance

rj-jesus commented:
Many thanks for the comments, @david-arm. I've tidied up the tests and added the checks for loop invariance (though I think they are redundant given how the IR is matched before, but in any case). Please let me know if that's what you had in mind and if you have any other comments. :)

rj-jesus commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
Thanks, done.

rj-jesus commented on llvm/test/CodeGen/AArch64/find-first-byte.ll:
```diff
@@ -0,0 +1,445 @@
+; DISABLE-NEXT:    [[TMP22:%.*]] = phi ptr [ [[TMP1]], [[TMP4:%.*]] ], [ [[DOTLCSSA]], %[[DOTLOOPEXIT]] ], [ [[TMP1]], %[[DOTLOOPEXIT1]] ]
+; DISABLE-NEXT:    ret ptr [[TMP22]]
+;
+  %5 = icmp eq ptr %0, %1
```
Thanks, please let me know if the new versions are better.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,432 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  // Add new incoming values with the result of the transformation to PHINodes
+  // of ExitSucc that use IndPhi.
+  for (auto *U : llvm::make_early_inc_range(IndPhi->users())) {
+    auto *PN = dyn_cast<PHINode>(U);
```
I think this code in recognize...

```
  for (User *U : IndPhi->users())
    if (!CurLoop->contains(cast<Instruction>(U))) {
      auto *PN = dyn_cast<PHINode>(U);
      if (!PN || PN->getParent() != ExitSucc)
        return false;
    }
```

guarantees that a) `U` is a PHINode and b) it lives in `ExitSucc`. Therefore, you should be able to assert this here with something like:

```
  auto *PN = cast<PHINode>(U);
  assert(PN->getParent() == ExitSucc && "User in unexpected block!");
  PN->addIncoming(MatchVal, cast<Instruction>(MatchVal)->getParent());
```

rj-jesus replied:
I think `IndPhi` also has users in the main loops, right? Namely the [load](https://github.com/llvm/llvm-project/blob/db2fbc69febbb073ab787385b78936c43c313105/llvm/test/Transforms/LoopIdiom/AArch64/find-first-byte.ll#L143) and the [gep](https://github.com/llvm/llvm-project/blob/db2fbc69febbb073ab787385b78936c43c313105/llvm/test/Transforms/LoopIdiom/AArch64/find-first-byte.ll#L158), I think, which are the users that `!CurLoop->contains(cast<Instruction>(U))` in the first snippet you pasted should be filtering out. I can add a similar check to `transformFindFirstByte` around the assert you suggested, do you think that would be better than the current check with `PN && PN->getParent() == ExitSucc`?

david-arm replied:
Sorry my mistake! It's because it doesn't have the `if (!CurLoop->contains(cast<Instruction>(U)))` guard so it could be inside the loop.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,432 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  if (!match(OuterBB->getTerminator(),
+             m_Br(m_SpecificICmp(ICmpInst::ICMP_EQ, m_Specific(GEPSearch),
+                                 m_Value(SearchEnd)),
+                  m_BasicBlock(ExitFail), m_Specific(Header))))
```
I didn't see any tests for the case when ExitFail != ExitSucc - can you add one please just to make sure it looks right?

rj-jesus replied:
Thanks, done in [`@find_first_of_i8_multi_exit`](https://github.com/llvm/llvm-project/pull/101976/commits/6d1755189549f3f6dd804d98810cbc374cd087a8#diff-f256205074dbd8200178875fe608f7daf7a0871a9e4bea54d8215524588aca76R325).

david-arm commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -939,3 +988,432 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
+  for (User *U : IndPhi->users())
+    if (!CurLoop->contains(cast<Instruction>(U))) {
+      auto *PN = dyn_cast<PHINode>(U);
+      if (!PN || PN->getParent() != ExitSucc)
```
Can you add a negative test for users in ExitFail (when ExitFail != ExitSucc)?

david-arm replied:
Yep, thanks!

david-arm reviewed (commented):
This looks great - thanks for being patient with all the review comments! I have a few more minor comments, mostly around adding a couple more tests, but then I think it's ready to go!

rj-jesus committed:
Add tests for different success/failure exits

rj-jesus commented on llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:
```diff
@@ -24,19 +26,42 @@
+//
+// NOTE: This Pass matches really specific loop patterns because it's only
+// supposed to be a temporary solution until our LoopVectorizer is powerful
+// enought to vectorize them automatically.
```
Done - thanks for catching this!

rj-jesus commented:
Thanks very much for the feedback, @david-arm! I've addressed most of your comments, please let me know if you'd like me to add anything else!

david-arm reviewed (approved):
LGTM!

rj-jesus commented:
Many thanks for the review! I'll let the tests run and merge it afterwards. :smile:

hiraditya commented:
Thanks for working on this. I wonder what other algorithms can benefit from the match sve2 instruction.

rj-jesus commented:
Hi @hiraditya, I think it can be useful for parsers and some compression algorithms. Do you have any particular use cases in mind?
(I see the tests I green, but I think I rather wait for next Monday to merge this in case any issues pop up!)

hiraditya commented:
> Hi @hiraditya, I think it can be useful for parsers and some compression algorithms. Do you have any particular use cases in mind? (I see the tests I green, but I think I rather wait for next Monday to merge this in case any issues pop up!)

no i was just curious. thanks.
